// react-pixel-display/dist/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("react/jsx-runtime");
const react = require("react");
const core_index = require("./index-H0p6sO8X.cjs");
function usePixelDisplay(options) {
  const {
    width = 64,
    height = 16,
    renderer: type = "imagedata",
    glow = true,
    scale = 8,
    pixelGap = 0.15
  } = options;
  const containerRef = react.useRef(null);
  const rendererRef = react.useRef(null);
  react.useEffect(() => {
    var _a;
    const container = containerRef.current;
    if (!container) return;
    (_a = rendererRef.current) == null ? void 0 : _a.destroy();
    const opts = { width, height, glow, scale, pixelGap };
    if (type === "canvas") {
      rendererRef.current = new core_index.CanvasLEDRenderer(container, opts);
    } else if (type === "svg") {
      rendererRef.current = new core_index.SVGLEDMatrixRenderer(container, { width, height, pixelGap });
    } else {
      rendererRef.current = new core_index.ImageDataLEDRenderer(container, opts);
    }
    return () => {
      var _a2;
      (_a2 = rendererRef.current) == null ? void 0 : _a2.destroy();
      rendererRef.current = null;
    };
  }, [width, height, type, glow, scale, pixelGap]);
  const setData = react.useCallback(
    (pixels, extended, extW) => {
      var _a;
      (_a = rendererRef.current) == null ? void 0 : _a.setData(pixels, extended, extW);
    },
    []
  );
  const setEffect = react.useCallback((effect, speed) => {
    var _a;
    (_a = rendererRef.current) == null ? void 0 : _a.setEffect(effect, speed);
  }, []);
  const start = react.useCallback(() => {
    var _a;
    return (_a = rendererRef.current) == null ? void 0 : _a.start();
  }, []);
  const stop = react.useCallback(() => {
    var _a;
    return (_a = rendererRef.current) == null ? void 0 : _a.stop();
  }, []);
  const renderStatic = react.useCallback(() => {
    var _a;
    return (_a = rendererRef.current) == null ? void 0 : _a.renderStatic();
  }, []);
  return { containerRef, rendererRef, setData, setEffect, start, stop, renderStatic };
}
function useTextRenderer(options) {
  const {
    width,
    height,
    text,
    font = "legacy",
    fontName = "VCR_OSD_MONO",
    foregroundColor = "#ff6600",
    backgroundColor = "#111",
    fontResolver,
    needsScroll = false
  } = options;
  const [fontReady, setFontReady] = react.useState(font === "legacy");
  react.useEffect(() => {
    if (font === "legacy") {
      setFontReady(true);
      return;
    }
    const heightKey = core_index.getHeightKey(height);
    if (font === "bdf") {
      if (core_index.isBdfFontLoaded(fontName, heightKey)) {
        setFontReady(true);
      } else {
        setFontReady(false);
        core_index.loadBdfFont(fontName, heightKey, fontResolver).then((result) => {
          setFontReady(!!result);
        });
      }
    } else {
      if (core_index.isFontLoaded(fontName)) {
        setFontReady(true);
      } else {
        setFontReady(false);
        core_index.loadFont(fontName, fontResolver).then((result) => {
          setFontReady(result);
        });
      }
    }
  }, [font, fontName, height, fontResolver]);
  const { pixels, scrollData } = react.useMemo(() => {
    if (!fontReady) return { pixels: null, scrollData: null };
    let result = null;
    if (font === "bdf") {
      result = core_index.textToPixelsBdf(text, width, height, foregroundColor, backgroundColor, fontName);
    } else if (font === "canvas") {
      result = core_index.textToPixelsCanvas(text, width, height, foregroundColor, backgroundColor, fontName);
    }
    if (!result) {
      result = core_index.textToPixels(text, width, height, foregroundColor, backgroundColor);
    }
    let scroll = null;
    if (needsScroll) {
      if (font === "bdf") {
        scroll = core_index.textToScrollPixelsBdf(text, width, height, foregroundColor, backgroundColor, fontName);
      } else if (font === "canvas") {
        scroll = core_index.textToScrollPixelsCanvas(text, width, height, foregroundColor, backgroundColor, fontName);
      }
      if (!scroll) {
        scroll = core_index.textToScrollPixels(text, width, height, foregroundColor, backgroundColor);
      }
    }
    return { pixels: result, scrollData: scroll };
  }, [fontReady, width, height, text, font, fontName, foregroundColor, backgroundColor, needsScroll]);
  return { pixels, scrollData, fontReady };
}
const PixelDisplay = react.forwardRef(
  (props, ref) => {
    var _a;
    const {
      width = 64,
      height = 16,
      pixels: propPixels,
      text,
      font = "legacy",
      fontName = "VCR_OSD_MONO",
      foregroundColor = "#ff6600",
      backgroundColor = "#111",
      effect = "fixed",
      speed = 50,
      renderer: rendererType = "imagedata",
      glow = true,
      scale = 8,
      pixelGap = 0.15,
      fontResolver,
      onReady,
      className,
      style
    } = props;
    const isAmbient = ((_a = core_index.EFFECTS[effect]) == null ? void 0 : _a.category) === "ambient";
    const needsScroll = (effect === "scroll_ltr" || effect === "scroll_rtl" || effect === "bounce") && !!text && !isAmbient;
    const { containerRef, rendererRef, setData, setEffect, start, stop, renderStatic } = usePixelDisplay({ width, height, renderer: rendererType, glow, scale, pixelGap });
    const textOptions = react.useMemo(
      () => ({
        width,
        height,
        text: text || "",
        font,
        fontName,
        foregroundColor,
        backgroundColor,
        fontResolver,
        needsScroll
      }),
      [width, height, text, font, fontName, foregroundColor, backgroundColor, fontResolver, needsScroll]
    );
    const { pixels: textPixels, scrollData, fontReady } = useTextRenderer(textOptions);
    react.useEffect(() => {
      if (!rendererRef.current) return;
      if (isAmbient) {
        setData([], [], width);
      } else if (propPixels) {
        setData(propPixels);
      } else if (textPixels) {
        if (needsScroll && scrollData) {
          setData(textPixels, scrollData.pixels, scrollData.width);
        } else {
          setData(textPixels);
        }
      }
    }, [propPixels, textPixels, scrollData, isAmbient, needsScroll, width, setData, rendererRef]);
    react.useEffect(() => {
      if (!rendererRef.current) return;
      setEffect(effect, speed);
      if (effect === "fixed") {
        stop();
        renderStatic();
      } else {
        start();
      }
    }, [effect, speed, setEffect, start, stop, renderStatic, rendererRef]);
    react.useImperativeHandle(
      ref,
      () => ({
        start: () => {
          var _a2;
          return (_a2 = rendererRef.current) == null ? void 0 : _a2.start();
        },
        stop: () => {
          var _a2;
          return (_a2 = rendererRef.current) == null ? void 0 : _a2.stop();
        },
        getRenderer: () => rendererRef.current,
        isRunning: () => {
          var _a2;
          return ((_a2 = rendererRef.current) == null ? void 0 : _a2.isRunning) ?? false;
        }
      }),
      [rendererRef]
    );
    react.useEffect(() => {
      if (rendererRef.current && (propPixels || textPixels || isAmbient)) {
        onReady == null ? void 0 : onReady();
      }
    }, [rendererRef, propPixels, textPixels, isAmbient, onReady]);
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        ref: containerRef,
        className,
        style: {
          background: "#0a0a0a",
          borderRadius: "4px",
          overflow: "hidden",
          ...style
        }
      }
    );
  }
);
PixelDisplay.displayName = "PixelDisplay";
exports.ALL_EFFECTS = core_index.ALL_EFFECTS;
exports.AMBIENT_EFFECTS = core_index.AMBIENT_EFFECTS;
exports.AmbientEffects = core_index.AmbientEffects;
exports.BDF_FONT_CONFIG = core_index.BDF_FONT_CONFIG;
exports.COLOR_EFFECTS = core_index.COLOR_EFFECTS;
exports.CanvasLEDRenderer = core_index.CanvasLEDRenderer;
exports.ColorEffects = core_index.ColorEffects;
exports.EFFECTS = core_index.EFFECTS;
exports.EFFECT_CATEGORIES = core_index.EFFECT_CATEGORIES;
exports.EffectManager = core_index.EffectManager;
exports.FONT_METRICS = core_index.FONT_METRICS;
exports.ImageDataLEDRenderer = core_index.ImageDataLEDRenderer;
exports.LEDMatrixRenderer = core_index.ImageDataLEDRenderer;
exports.SVGLEDMatrixRenderer = core_index.SVGLEDMatrixRenderer;
exports.TEXT_EFFECTS = core_index.TEXT_EFFECTS;
exports.TextEffects = core_index.TextEffects;
exports.configureFonts = core_index.configureFonts;
exports.createPixelSvg = core_index.createPixelSvg;
exports.getAvailableBdfFonts = core_index.getAvailableBdfFonts;
exports.getHeightKey = core_index.getHeightKey$1;
exports.hexToRgb = core_index.hexToRgb;
exports.hexToRgbObj = core_index.hexToRgbObj;
exports.hsvToRgb = core_index.hsvToRgb;
exports.isBdfFontLoaded = core_index.isBdfFontLoaded;
exports.isFontLoaded = core_index.isFontLoaded;
exports.loadBdfFont = core_index.loadBdfFont;
exports.loadFont = core_index.loadFont;
exports.pixelFont = core_index.pixelFont;
exports.preloadBdfFonts = core_index.preloadBdfFonts;
exports.preloadFonts = core_index.preloadFonts;
exports.setBdfFontResolver = core_index.setFontResolver;
exports.setTtfFontResolver = core_index.setFontResolver$1;
exports.textToPixels = core_index.textToPixels;
exports.textToPixelsBdf = core_index.textToPixelsBdf;
exports.textToPixelsCanvas = core_index.textToPixelsCanvas;
exports.textToScrollPixels = core_index.textToScrollPixels;
exports.textToScrollPixelsBdf = core_index.textToScrollPixelsBdf;
exports.textToScrollPixelsCanvas = core_index.textToScrollPixelsCanvas;
exports.PixelDisplay = PixelDisplay;
exports.usePixelDisplay = usePixelDisplay;
exports.useTextRenderer = useTextRenderer;
