// react-pixel-display/dist/index-DWEf46Kx.js
function hexToRgb(hex) {
  if (!hex || hex === "#111" || hex === "#000") return [17, 17, 17];
  if (hex === "#050505") return [5, 5, 5];
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result ? [parseInt(result[1], 16), parseInt(result[2], 16), parseInt(result[3], 16)] : [17, 17, 17];
}
function hexToRgbObj(hex) {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result ? {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16)
  } : { r: 0, g: 0, b: 0 };
}
function hsvToRgb(h, s, v) {
  let r = 0, g = 0, b = 0;
  const i = Math.floor(h * 6);
  const f = h * 6 - i;
  const p = v * (1 - s);
  const q = v * (1 - f * s);
  const t = v * (1 - (1 - f) * s);
  switch (i % 6) {
    case 0:
      r = v;
      g = t;
      b = p;
      break;
    case 1:
      r = q;
      g = v;
      b = p;
      break;
    case 2:
      r = p;
      g = v;
      b = t;
      break;
    case 3:
      r = p;
      g = q;
      b = v;
      break;
    case 4:
      r = t;
      g = p;
      b = v;
      break;
    case 5:
      r = v;
      g = p;
      b = q;
      break;
  }
  return [r * 255, g * 255, b * 255];
}
class TextEffects {
  constructor(renderer) {
    this.renderer = renderer;
  }
  init(effectName, state) {
    const { width, height } = this.renderer;
    switch (effectName) {
      case "scroll_ltr":
      case "scroll_rtl":
        state.offset = 0;
        break;
      case "blink":
        state.visible = true;
        break;
      case "snow":
      case "breeze":
        state.phases = [];
        for (let i = 0; i < width * height; i++) {
          state.phases.push(Math.random() * Math.PI * 2);
        }
        break;
      case "laser":
        state.position = 0;
        break;
      case "fade":
        state.opacity = 0;
        state.direction = 1;
        break;
      case "typewriter":
        state.charIndex = 0;
        state.cursorVisible = true;
        break;
      case "bounce":
        state.offset = 0;
        state.direction = 1;
        break;
      case "sparkle":
        state.sparkles = [];
        for (let i = 0; i < Math.floor(width * height * 0.1); i++) {
          state.sparkles.push({
            x: Math.floor(Math.random() * width),
            y: Math.floor(Math.random() * height),
            brightness: Math.random(),
            speed: 0.05 + Math.random() * 0.1
          });
        }
        break;
    }
  }
  step(effectName, state) {
    const { width, extendedWidth } = this.renderer;
    switch (effectName) {
      case "scroll_ltr":
        state.offset -= 1;
        if (state.offset <= -(extendedWidth || width)) {
          state.offset = width;
        }
        break;
      case "scroll_rtl":
        state.offset += 1;
        if (state.offset >= (extendedWidth || width)) {
          state.offset = -width;
        }
        break;
      case "blink":
        state.visible = !state.visible;
        break;
      case "laser":
        state.position = (state.position + 1) % width;
        break;
      case "fade":
        state.opacity += state.direction * 0.05;
        if (state.opacity >= 1) {
          state.opacity = 1;
          state.direction = -1;
        } else if (state.opacity <= 0) {
          state.opacity = 0;
          state.direction = 1;
        }
        break;
      case "typewriter":
        if (state.tick % 3 === 0) {
          state.charIndex++;
        }
        state.cursorVisible = state.tick % 10 < 5;
        break;
      case "bounce": {
        state.offset += state.direction;
        const maxOffset = Math.max(0, (extendedWidth || width) - width);
        if (state.offset >= maxOffset) {
          state.offset = maxOffset;
          state.direction = -1;
        } else if (state.offset <= 0) {
          state.offset = 0;
          state.direction = 1;
        }
        break;
      }
      case "sparkle": {
        const sparkles = state.sparkles;
        for (const sparkle of sparkles) {
          sparkle.brightness += sparkle.speed;
          if (sparkle.brightness > 1) {
            sparkle.brightness = 0;
            sparkle.x = Math.floor(Math.random() * width);
            sparkle.y = Math.floor(Math.random() * this.renderer.height);
          }
        }
        break;
      }
    }
  }
  render(effectName, state, pixels, extendedPixels, extendedWidth) {
    const { width, height } = this.renderer;
    const srcPixels = extendedPixels || pixels || [];
    const displayPixels = pixels || [];
    const srcWidth = extendedWidth || width;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        let color;
        let sourceX = x;
        if (effectName === "scroll_ltr" || effectName === "scroll_rtl" || effectName === "bounce") {
          sourceX = x - (state.offset || 0);
          while (sourceX < 0) sourceX += srcWidth;
          while (sourceX >= srcWidth) sourceX -= srcWidth;
          color = srcPixels[y * srcWidth + sourceX] || "#111";
        } else if (effectName === "typewriter") {
          const charWidth = 6;
          const maxX = (state.charIndex || 0) * charWidth;
          if (x < maxX) {
            color = displayPixels[y * width + x] || "#111";
          } else if (x === maxX && state.cursorVisible) {
            color = "#ffffff";
          } else {
            color = "#111";
          }
        } else {
          color = displayPixels[y * width + x] || "#111";
        }
        let [r, g, b] = hexToRgb(color);
        const isLit = r > 20 || g > 20 || b > 20;
        if (isLit) {
          switch (effectName) {
            case "blink":
              if (!state.visible) {
                r = g = b = 17;
              }
              break;
            case "snow": {
              const phases = state.phases;
              const phase = (phases == null ? void 0 : phases[y * width + x]) || 0;
              const tick = state.tick || 0;
              const factor = 0.3 + 0.7 * Math.abs(Math.sin(phase + tick * 0.3));
              r *= factor;
              g *= factor;
              b *= factor;
              break;
            }
            case "breeze": {
              const phases = state.phases;
              const phase = (phases == null ? void 0 : phases[y * width + x]) || 0;
              const tick = state.tick || 0;
              const factor = 0.4 + 0.6 * Math.abs(Math.sin(phase + tick * 0.15 + x * 0.2));
              r *= factor;
              g *= factor;
              b *= factor;
              break;
            }
            case "laser": {
              const pos = state.position || 0;
              const dist = Math.abs(x - pos);
              const factor = dist < 3 ? 1 : 0.3;
              r *= factor;
              g *= factor;
              b *= factor;
              break;
            }
            case "fade": {
              const opacity = state.opacity || 1;
              r *= opacity;
              g *= opacity;
              b *= opacity;
              break;
            }
          }
        }
        if (effectName === "sparkle" && state.sparkles) {
          const sparkles = state.sparkles;
          for (const sparkle of sparkles) {
            if (sparkle.x === x && sparkle.y === y) {
              const sparkleIntensity = Math.sin(sparkle.brightness * Math.PI);
              r = Math.min(255, r + sparkleIntensity * 200);
              g = Math.min(255, g + sparkleIntensity * 200);
              b = Math.min(255, b + sparkleIntensity * 200);
            }
          }
        }
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
}
class AmbientEffects {
  constructor(renderer) {
    this.renderer = renderer;
  }
  init(effectName, state) {
    const { width, height } = this.renderer;
    switch (effectName) {
      case "rainbow":
        state.position = 0;
        break;
      case "matrix": {
        const colorModes = [
          [0, 255, 0],
          [0, 255, 255],
          [255, 0, 255]
        ];
        state.colorMode = colorModes[Math.floor(Math.random() * colorModes.length)];
        state.buffer = [];
        for (let y = 0; y < height; y++) {
          state.buffer.push(
            Array(width).fill(null).map(() => [0, 0, 0])
          );
        }
        break;
      }
      case "plasma":
      case "gradient":
        state.time = 0;
        break;
      case "fire":
        state.heat = [];
        for (let i = 0; i < width * height; i++) {
          state.heat.push(0);
        }
        state.palette = this._createFirePalette();
        break;
      case "water":
        state.current = [];
        state.previous = [];
        for (let i = 0; i < width * height; i++) {
          state.current.push(0);
          state.previous.push(0);
        }
        state.damping = 0.95;
        break;
      case "stars": {
        state.stars = [];
        const numStars = Math.floor(width * height * 0.15);
        for (let i = 0; i < numStars; i++) {
          state.stars.push({
            x: Math.floor(Math.random() * width),
            y: Math.floor(Math.random() * height),
            brightness: Math.random(),
            speed: 0.02 + Math.random() * 0.05,
            phase: Math.random() * Math.PI * 2
          });
        }
        break;
      }
      case "confetti":
        state.particles = [];
        for (let i = 0; i < 20; i++) {
          state.particles.push(this._createConfettiParticle(width, height, true));
        }
        break;
      case "plasma_wave":
      case "radial_pulse":
      case "hypnotic":
      case "aurora":
        state.time = 0;
        break;
      case "lava":
        state.time = 0;
        state.noise = [];
        for (let i = 0; i < width * height; i++) {
          state.noise.push(Math.random() * Math.PI * 2);
        }
        break;
    }
  }
  step(effectName, state) {
    const { width, height } = this.renderer;
    switch (effectName) {
      case "rainbow":
        state.position = (state.position + 0.01) % 1;
        break;
      case "matrix":
        this._stepMatrix(state, width, height);
        break;
      case "plasma":
      case "gradient":
        state.time = (state.time || 0) + 0.05;
        break;
      case "fire":
        this._stepFire(state, width, height);
        break;
      case "water":
        this._stepWater(state, width, height);
        break;
      case "stars": {
        const stars = state.stars;
        for (const star of stars) {
          star.phase += star.speed;
        }
        break;
      }
      case "confetti": {
        const particles = state.particles;
        for (let i = 0; i < particles.length; i++) {
          const p = particles[i];
          p.y += p.speed;
          p.x += p.drift;
          p.rotation += p.rotationSpeed;
          if (p.y > height) {
            particles[i] = this._createConfettiParticle(width, height, false);
          }
        }
        break;
      }
      case "plasma_wave":
      case "radial_pulse":
      case "hypnotic":
      case "lava":
      case "aurora":
        state.time = (state.time || 0) + 0.03;
        break;
    }
  }
  render(effectName, state) {
    switch (effectName) {
      case "rainbow":
        this._renderRainbow(state);
        break;
      case "matrix":
        this._renderMatrix(state);
        break;
      case "plasma":
        this._renderPlasma(state);
        break;
      case "gradient":
        this._renderGradient(state);
        break;
      case "fire":
        this._renderFire(state);
        break;
      case "water":
        this._renderWater(state);
        break;
      case "stars":
        this._renderStars(state);
        break;
      case "confetti":
        this._renderConfetti(state);
        break;
      case "plasma_wave":
        this._renderPlasmaWave(state);
        break;
      case "radial_pulse":
        this._renderRadialPulse(state);
        break;
      case "hypnotic":
        this._renderHypnotic(state);
        break;
      case "lava":
        this._renderLava(state);
        break;
      case "aurora":
        this._renderAurora(state);
        break;
    }
  }
  _renderRainbow(state) {
    const { width, height } = this.renderer;
    const position = state.position || 0;
    for (let x = 0; x < width; x++) {
      const hue = (position + x / width) % 1;
      const [r, g, b] = hsvToRgb(hue, 1, 0.6);
      for (let y = 0; y < height; y++) {
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _stepMatrix(state, width, height) {
    const buffer = state.buffer;
    const colorMode = state.colorMode;
    const fadeAmount = 0.15;
    buffer.pop();
    const newRow = buffer[0].map(([r, g, b]) => [
      r * (1 - fadeAmount),
      g * (1 - fadeAmount),
      b * (1 - fadeAmount)
    ]);
    buffer.unshift(JSON.parse(JSON.stringify(newRow)));
    for (let x = 0; x < width; x++) {
      if (Math.random() < 0.08) {
        buffer[0][x] = [
          Math.floor(Math.random() * colorMode[0]),
          Math.floor(Math.random() * colorMode[1]),
          Math.floor(Math.random() * colorMode[2])
        ];
      }
    }
  }
  _renderMatrix(state) {
    var _a;
    const { width, height } = this.renderer;
    const buffer = state.buffer;
    if (!buffer) return;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const [r, g, b] = ((_a = buffer[y]) == null ? void 0 : _a[x]) || [0, 0, 0];
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _renderPlasma(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    const centerX = width / 2;
    const centerY = height / 2;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const dx = x - centerX;
        const dy = y - centerY;
        const dist = Math.sqrt(dx * dx + dy * dy);
        const v1 = Math.sin(x / 8 + time);
        const v2 = Math.sin(y / 6 + time * 0.8);
        const v3 = Math.sin(dist / 6 - time * 1.2);
        const v4 = Math.sin((x + y) / 10 + time * 0.5);
        const value = (v1 + v2 + v3 + v4 + 4) / 8;
        const r = Math.sin(value * Math.PI * 2) * 0.5 + 0.5;
        const g = Math.sin(value * Math.PI * 2 + 2) * 0.5 + 0.5;
        const b = Math.sin(value * Math.PI * 2 + 4) * 0.5 + 0.5;
        this.renderer.setPixel(x, y, [r * 255, g * 255, b * 255]);
      }
    }
  }
  _renderGradient(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    const t = time * 10;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const r = (Math.sin((x + t) * 0.05) * 0.5 + 0.5) * 255;
        const g = (Math.cos((y + t) * 0.05) * 0.5 + 0.5) * 255;
        const b = (Math.sin((x + y + t) * 0.03) * 0.5 + 0.5) * 255;
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _createFirePalette() {
    const palette = [];
    for (let i = 0; i < 256; i++) {
      let r, g, b;
      if (i < 64) {
        r = i * 4;
        g = 0;
        b = 0;
      } else if (i < 128) {
        r = 255;
        g = (i - 64) * 4;
        b = 0;
      } else if (i < 192) {
        r = 255;
        g = 255;
        b = (i - 128) * 4;
      } else {
        r = 255;
        g = 255;
        b = 255;
      }
      palette.push([r, g, b]);
    }
    return palette;
  }
  _stepFire(state, width, height) {
    const heat = state.heat;
    for (let i = 0; i < width * height; i++) {
      heat[i] = Math.max(0, heat[i] - Math.random() * 10);
    }
    for (let y = 0; y < height - 1; y++) {
      for (let x = 0; x < width; x++) {
        const idx = y * width + x;
        const below = (y + 1) * width + x;
        const left = y * width + Math.max(0, x - 1);
        const right = y * width + Math.min(width - 1, x + 1);
        heat[idx] = (heat[below] + heat[left] + heat[right]) / 3.05;
      }
    }
    for (let x = 0; x < width; x++) {
      if (Math.random() < 0.6) {
        heat[(height - 1) * width + x] = 180 + Math.random() * 75;
      }
    }
  }
  _renderFire(state) {
    const { width, height } = this.renderer;
    const heat = state.heat;
    const palette = state.palette;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const idx = y * width + x;
        const h = Math.floor(Math.min(255, heat[idx]));
        const [r, g, b] = palette[h];
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _stepWater(state, width, height) {
    const current = state.current;
    const previous = state.previous;
    const damping = state.damping;
    const temp = [...previous];
    for (let i = 0; i < current.length; i++) {
      previous[i] = current[i];
    }
    for (let y = 1; y < height - 1; y++) {
      for (let x = 1; x < width - 1; x++) {
        const idx = y * width + x;
        current[idx] = (temp[(y - 1) * width + x] + temp[(y + 1) * width + x] + temp[y * width + (x - 1)] + temp[y * width + (x + 1)]) / 2 - current[idx];
        current[idx] *= damping;
      }
    }
    if (Math.random() < 0.1) {
      const x = Math.floor(Math.random() * (width - 2)) + 1;
      const y = Math.floor(Math.random() * (height - 2)) + 1;
      current[y * width + x] = 255;
    }
  }
  _renderWater(state) {
    const { width, height } = this.renderer;
    const current = state.current;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const idx = y * width + x;
        const value = Math.abs(current[idx]);
        const intensity = Math.min(255, value * 2);
        const r = intensity > 200 ? intensity : 0;
        const g = intensity > 150 ? intensity * 0.8 : intensity * 0.3;
        const b = Math.min(255, 50 + intensity);
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _renderStars(state) {
    const { width, height } = this.renderer;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        this.renderer.setPixel(x, y, [5, 5, 15]);
      }
    }
    const stars = state.stars;
    for (const star of stars) {
      const brightness = (Math.sin(star.phase) * 0.5 + 0.5) * 255;
      const x = Math.floor(star.x);
      const y = Math.floor(star.y);
      if (x >= 0 && x < width && y >= 0 && y < height) {
        this.renderer.setPixel(x, y, [brightness, brightness, brightness * 0.9]);
      }
    }
  }
  _createConfettiParticle(width, height, randomY) {
    const colors = [
      [255, 0, 0],
      [0, 255, 0],
      [0, 0, 255],
      [255, 255, 0],
      [255, 0, 255],
      [0, 255, 255],
      [255, 128, 0],
      [255, 192, 203]
    ];
    return {
      x: Math.random() * width,
      y: randomY ? Math.random() * height : -2,
      speed: 0.2 + Math.random() * 0.3,
      drift: (Math.random() - 0.5) * 0.3,
      color: colors[Math.floor(Math.random() * colors.length)],
      size: 1 + Math.random(),
      rotation: Math.random() * Math.PI * 2,
      rotationSpeed: (Math.random() - 0.5) * 0.2
    };
  }
  _renderConfetti(state) {
    const { width, height } = this.renderer;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        this.renderer.setPixel(x, y, [10, 10, 10]);
      }
    }
    const particles = state.particles;
    for (const p of particles) {
      const x = Math.floor(p.x);
      const y = Math.floor(p.y);
      if (x >= 0 && x < width && y >= 0 && y < height) {
        const shimmer = Math.abs(Math.sin(p.rotation)) * 0.5 + 0.5;
        const [r, g, b] = p.color;
        this.renderer.setPixel(x, y, [r * shimmer, g * shimmer, b * shimmer]);
      }
    }
  }
  _renderPlasmaWave(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const uvX = x / width;
        const uvY = y / height;
        const v = Math.sin(uvX * 10 + time) + Math.sin(uvY * 10 + time) + Math.sin((uvX + uvY) * 10 + time) + Math.sin(Math.sqrt((uvX - 0.5) ** 2 + (uvY - 0.5) ** 2) * 20 - time * 2);
        const r = Math.sin(v * Math.PI) * 0.5 + 0.5;
        const g = Math.sin(v * Math.PI + 2.094) * 0.5 + 0.5;
        const b = Math.sin(v * Math.PI + 4.188) * 0.5 + 0.5;
        this.renderer.setPixel(x, y, [r * 255, g * 255, b * 255]);
      }
    }
  }
  _renderRadialPulse(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    const centerX = width / 2;
    const centerY = height / 2;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const dx = x - centerX;
        const dy = y - centerY;
        const dist = Math.sqrt(dx * dx + dy * dy);
        const wave = Math.sin(dist * 0.8 - time * 3) * 0.5 + 0.5;
        const pulse = Math.sin(time * 2) * 0.3 + 0.7;
        const hue = (dist / 20 + time * 0.5) % 1;
        const [r, g, b] = hsvToRgb(hue, 0.8, wave * pulse);
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _renderHypnotic(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    const centerX = width / 2;
    const centerY = height / 2;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const dx = x - centerX;
        const dy = y - centerY;
        const dist = Math.sqrt(dx * dx + dy * dy);
        const angle = Math.atan2(dy, dx);
        const spiral = Math.sin(angle * 4 + dist * 0.5 - time * 2);
        const intensity = spiral * 0.5 + 0.5;
        const r = intensity * (Math.sin(time) * 0.5 + 0.5);
        const g = intensity * (Math.sin(time + 2.094) * 0.5 + 0.5);
        const b = intensity * (Math.sin(time + 4.188) * 0.5 + 0.5);
        this.renderer.setPixel(x, y, [r * 255, g * 255, b * 255]);
      }
    }
  }
  _renderLava(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const uvX = x / width;
        const uvY = y / height;
        const n1 = Math.sin(uvX * 8 + time * 0.7) * Math.cos(uvY * 6 + time * 0.5);
        const n2 = Math.sin(uvX * 12 - time * 0.3) * Math.sin(uvY * 10 + time * 0.8);
        const n3 = Math.cos((uvX + uvY) * 5 + time);
        const value = (n1 + n2 + n3 + 3) / 6;
        let r, g, b;
        if (value < 0.3) {
          r = value * 3 * 100;
          g = 0;
          b = 0;
        } else if (value < 0.6) {
          r = 100 + (value - 0.3) * 3 * 155;
          g = (value - 0.3) * 3 * 100;
          b = 0;
        } else {
          r = 255;
          g = 100 + (value - 0.6) * 2.5 * 155;
          b = (value - 0.6) * 2.5 * 100;
        }
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
  _renderAurora(state) {
    const { width, height } = this.renderer;
    const time = state.time || 0;
    for (let x = 0; x < width; x++) {
      for (let y = 0; y < height; y++) {
        const uvX = x / width;
        const uvY = y / height;
        const wave1 = Math.sin(uvX * 6 + time) * 0.3;
        const wave2 = Math.sin(uvX * 4 - time * 0.7) * 0.2;
        const wave3 = Math.sin(uvX * 8 + time * 1.3) * 0.15;
        const waveLine = 0.5 + wave1 + wave2 + wave3;
        const distFromWave = Math.abs(uvY - waveLine);
        const intensity = Math.max(0, 1 - distFromWave * 4);
        const glow = Math.pow(intensity, 1.5);
        const colorShift = Math.sin(uvX * 3 + time * 0.5);
        let r = glow * (0.2 + colorShift * 0.3) * 255;
        let g = glow * (0.8 + Math.sin(time + uvX) * 0.2) * 255;
        let b = glow * (0.6 + colorShift * 0.4) * 255;
        const starChance = Math.sin(x * 127.1 + y * 311.7) * 0.5 + 0.5;
        const starTwinkle = Math.sin(time * 3 + x + y) * 0.5 + 0.5;
        if (starChance > 0.98 && intensity < 0.3) {
          const starBright = starTwinkle * 180;
          r = Math.max(r, starBright);
          g = Math.max(g, starBright);
          b = Math.max(b, starBright * 0.9);
        }
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
}
class ColorEffects {
  constructor(renderer) {
    this.renderer = renderer;
  }
  init(effectName, state) {
    switch (effectName) {
      case "color_cycle":
        state.hue = 0;
        break;
      case "rainbow_text":
        state.offset = 0;
        break;
      case "neon":
        state.glowIntensity = 0;
        state.direction = 1;
        state.baseColor = state.fgColor || "#ff00ff";
        break;
    }
  }
  step(effectName, state) {
    switch (effectName) {
      case "color_cycle":
        state.hue = (state.hue + 0.01) % 1;
        break;
      case "rainbow_text":
        state.offset = (state.offset + 0.02) % 1;
        break;
      case "neon":
        state.glowIntensity += state.direction * 0.05;
        if (state.glowIntensity >= 1) {
          state.glowIntensity = 1;
          state.direction = -1;
        } else if (state.glowIntensity <= 0.3) {
          state.glowIntensity = 0.3;
          state.direction = 1;
        }
        break;
    }
  }
  render(effectName, state, pixels) {
    const { width, height } = this.renderer;
    const displayPixels = pixels || [];
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const color = displayPixels[y * width + x] || "#111";
        let [r, g, b] = hexToRgb(color);
        const isLit = r > 20 || g > 20 || b > 20;
        if (isLit) {
          switch (effectName) {
            case "color_cycle": {
              const [nr, ng, nb] = hsvToRgb(state.hue, 1, 0.8);
              const brightness = (r + g + b) / (3 * 255);
              r = nr * brightness;
              g = ng * brightness;
              b = nb * brightness;
              break;
            }
            case "rainbow_text": {
              const hue = (state.offset + x / width) % 1;
              const [nr, ng, nb] = hsvToRgb(hue, 1, 0.8);
              const brightness = (r + g + b) / (3 * 255);
              r = nr * brightness;
              g = ng * brightness;
              b = nb * brightness;
              break;
            }
            case "neon": {
              const baseColor = hexToRgb(state.baseColor || "#ff00ff");
              const intensity = state.glowIntensity || 0.5;
              r = baseColor[0] * intensity;
              g = baseColor[1] * intensity;
              b = baseColor[2] * intensity;
              if (intensity > 0.8) {
                const whiteMix = (intensity - 0.8) * 5;
                r = r + (255 - r) * whiteMix * 0.3;
                g = g + (255 - g) * whiteMix * 0.3;
                b = b + (255 - b) * whiteMix * 0.3;
              }
              break;
            }
          }
        }
        this.renderer.setPixel(x, y, [r, g, b]);
      }
    }
  }
}
const EFFECT_CATEGORIES = {
  TEXT: "text",
  AMBIENT: "ambient",
  COLOR: "color"
};
const EFFECTS = {
  // Text effects
  fixed: { category: "text", name: "Fixed", description: "Static display" },
  scroll_ltr: { category: "text", name: "Scroll Left", description: "Text scrolls left to right" },
  scroll_rtl: { category: "text", name: "Scroll Right", description: "Text scrolls right to left" },
  blink: { category: "text", name: "Blink", description: "Text blinks on/off" },
  breeze: { category: "text", name: "Breeze", description: "Gentle wave brightness" },
  snow: { category: "text", name: "Snow", description: "Sparkle effect" },
  laser: { category: "text", name: "Laser", description: "Scanning beam" },
  fade: { category: "text", name: "Fade", description: "Fade in/out" },
  typewriter: { category: "text", name: "Typewriter", description: "Characters appear one by one" },
  bounce: { category: "text", name: "Bounce", description: "Text bounces back and forth" },
  sparkle: { category: "text", name: "Sparkle", description: "Random sparkle overlay" },
  // Ambient effects
  rainbow: { category: "ambient", name: "Rainbow", description: "HSV rainbow gradient" },
  matrix: { category: "ambient", name: "Matrix", description: "Digital rain effect" },
  plasma: { category: "ambient", name: "Plasma", description: "Classic plasma waves" },
  gradient: { category: "ambient", name: "Gradient", description: "Moving color gradients" },
  fire: { category: "ambient", name: "Fire", description: "Fire/flame simulation" },
  water: { category: "ambient", name: "Water", description: "Ripple/wave effect" },
  stars: { category: "ambient", name: "Stars", description: "Twinkling starfield" },
  confetti: { category: "ambient", name: "Confetti", description: "Falling colored particles" },
  plasma_wave: { category: "ambient", name: "Plasma Wave", description: "Multi-frequency sine waves" },
  radial_pulse: { category: "ambient", name: "Radial Pulse", description: "Expanding ring patterns" },
  hypnotic: { category: "ambient", name: "Hypnotic", description: "Spiral pattern" },
  lava: { category: "ambient", name: "Lava", description: "Flowing lava/magma" },
  aurora: { category: "ambient", name: "Aurora", description: "Northern lights" },
  // Color effects
  color_cycle: { category: "color", name: "Color Cycle", description: "Cycle through colors" },
  rainbow_text: { category: "color", name: "Rainbow Text", description: "Rainbow gradient on text" },
  neon: { category: "color", name: "Neon", description: "Pulsing neon glow" }
};
class EffectManager {
  constructor(renderer) {
    this.renderer = renderer;
    this.textEffects = new TextEffects(renderer);
    this.ambientEffects = new AmbientEffects(renderer);
    this.colorEffects = new ColorEffects(renderer);
    this.currentEffect = "fixed";
    this.effectState = { tick: 0 };
  }
  getEffectInfo(effectName) {
    return EFFECTS[effectName] || EFFECTS.fixed;
  }
  getEffectsByCategory(category) {
    return Object.entries(EFFECTS).filter(([, info]) => info.category === category).map(([key, info]) => ({ key, ...info }));
  }
  initEffect(effectName, options = {}) {
    const info = this.getEffectInfo(effectName);
    this.currentEffect = effectName;
    this.effectState = { tick: 0, ...options };
    switch (info.category) {
      case "text":
        this.textEffects.init(effectName, this.effectState);
        break;
      case "ambient":
        this.ambientEffects.init(effectName, this.effectState);
        break;
      case "color":
        this.colorEffects.init(effectName, this.effectState);
        break;
    }
    return this.effectState;
  }
  step() {
    const info = this.getEffectInfo(this.currentEffect);
    this.effectState.tick = (this.effectState.tick || 0) + 1;
    switch (info.category) {
      case "text":
        this.textEffects.step(this.currentEffect, this.effectState);
        break;
      case "ambient":
        this.ambientEffects.step(this.currentEffect, this.effectState);
        break;
      case "color":
        this.colorEffects.step(this.currentEffect, this.effectState);
        break;
    }
  }
  render(pixels, extendedPixels, extendedWidth) {
    const info = this.getEffectInfo(this.currentEffect);
    switch (info.category) {
      case "ambient":
        this.ambientEffects.render(this.currentEffect, this.effectState);
        break;
      case "text":
        this.textEffects.render(this.currentEffect, this.effectState, pixels, extendedPixels, extendedWidth);
        break;
      case "color":
        this.colorEffects.render(this.currentEffect, this.effectState, pixels);
        break;
    }
  }
  isAmbient(effectName) {
    return this.getEffectInfo(effectName).category === "ambient";
  }
  needsAnimation(effectName) {
    return effectName !== "fixed";
  }
}
const TEXT_EFFECTS = Object.entries(EFFECTS).filter(([, info]) => info.category === "text").map(([name]) => name);
const AMBIENT_EFFECTS = Object.entries(EFFECTS).filter(([, info]) => info.category === "ambient").map(([name]) => name);
const COLOR_EFFECTS = Object.entries(EFFECTS).filter(([, info]) => info.category === "color").map(([name]) => name);
const ALL_EFFECTS = Object.keys(EFFECTS);
class ImageDataLEDRenderer {
  constructor(container, options = {}) {
    this.container = container;
    this.width = options.width || 64;
    this.height = options.height || 16;
    this.pixelGap = options.pixelGap || 0.15;
    this.glowEnabled = options.glow !== false;
    this.scale = options.scale || 8;
    this.buffer = [];
    this._initBuffer();
    this._colorPixels = [];
    this._extendedColorPixels = [];
    this.extendedWidth = this.width;
    this.effect = "fixed";
    this.speed = 50;
    this.animationId = null;
    this.lastFrameTime = 0;
    this._isRunning = false;
    this._canvas = null;
    this._ctx = null;
    this._imageData = null;
    this._glowCanvas = null;
    this._glowCtx = null;
    this._wrapper = null;
    this._canvasCreated = false;
    this._pixelTemplate = null;
    this.effectManager = new EffectManager(this);
  }
  _initBuffer() {
    this.buffer = [];
    for (let i = 0; i < this.width * this.height; i++) {
      this.buffer.push([0, 0, 0]);
    }
  }
  _createCanvas() {
    if (typeof document === "undefined") return;
    const canvasWidth = this.width * this.scale;
    const canvasHeight = this.height * this.scale;
    this._wrapper = document.createElement("div");
    this._wrapper.style.cssText = `
      position: relative;
      width: 100%;
      aspect-ratio: ${this.width} / ${this.height};
      background: #0a0a0a;
      border-radius: 4px;
      overflow: hidden;
    `;
    if (this.glowEnabled) {
      this._glowCanvas = document.createElement("canvas");
      this._glowCanvas.width = canvasWidth;
      this._glowCanvas.height = canvasHeight;
      this._glowCanvas.style.cssText = `
        position: absolute; top: 0; left: 0; width: 100%; height: 100%;
        filter: blur(${this.scale * 0.6}px); opacity: 0.5;
      `;
      this._glowCtx = this._glowCanvas.getContext("2d", { alpha: false });
      this._wrapper.appendChild(this._glowCanvas);
    }
    this._canvas = document.createElement("canvas");
    this._canvas.width = canvasWidth;
    this._canvas.height = canvasHeight;
    this._canvas.style.cssText = `
      position: absolute; top: 0; left: 0; width: 100%; height: 100%;
      image-rendering: pixelated; image-rendering: crisp-edges;
    `;
    this._ctx = this._canvas.getContext("2d", { alpha: false });
    this._wrapper.appendChild(this._canvas);
    this._imageData = this._ctx.createImageData(canvasWidth, canvasHeight);
    this._createPixelTemplate();
    this._fillBackground();
    if (this.container && this.container.isConnected !== false) {
      this.container.innerHTML = "";
      this.container.appendChild(this._wrapper);
    }
    this._canvasCreated = true;
  }
  _createPixelTemplate() {
    const scale = this.scale;
    const gap = Math.max(1, Math.floor(scale * this.pixelGap));
    const pixelSize = scale - gap;
    const radius = Math.max(1, Math.floor(scale * 0.15));
    this._pixelTemplate = [];
    for (let py = 0; py < scale; py++) {
      for (let px = 0; px < scale; px++) {
        let inside = false;
        if (px < pixelSize && py < pixelSize) {
          if (px < radius && py < radius) {
            const dx = radius - px;
            const dy = radius - py;
            inside = dx * dx + dy * dy <= radius * radius;
          } else if (px >= pixelSize - radius && py < radius) {
            const dx = px - (pixelSize - radius - 1);
            const dy = radius - py;
            inside = dx * dx + dy * dy <= radius * radius;
          } else if (px < radius && py >= pixelSize - radius) {
            const dx = radius - px;
            const dy = py - (pixelSize - radius - 1);
            inside = dx * dx + dy * dy <= radius * radius;
          } else if (px >= pixelSize - radius && py >= pixelSize - radius) {
            const dx = px - (pixelSize - radius - 1);
            const dy = py - (pixelSize - radius - 1);
            inside = dx * dx + dy * dy <= radius * radius;
          } else {
            inside = true;
          }
        }
        this._pixelTemplate.push(inside);
      }
    }
  }
  _fillBackground() {
    if (!this._imageData) return;
    const data = this._imageData.data;
    const bgR = 10, bgG = 10, bgB = 10;
    for (let i = 0; i < data.length; i += 4) {
      data[i] = bgR;
      data[i + 1] = bgG;
      data[i + 2] = bgB;
      data[i + 3] = 255;
    }
  }
  _ensureCanvasInContainer() {
    if (!this.container) return false;
    if (this._wrapper && this._wrapper.parentNode === this.container) return true;
    if (this._wrapper && this.container.isConnected !== false) {
      this.container.innerHTML = "";
      this.container.appendChild(this._wrapper);
      return true;
    }
    return false;
  }
  setPixel(x, y, color) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      const idx = y * this.width + x;
      if (idx < this.buffer.length) {
        this.buffer[idx] = color;
      }
    }
  }
  clear() {
    for (let i = 0; i < this.buffer.length; i++) {
      this.buffer[i] = [0, 0, 0];
    }
  }
  flush() {
    if (!this._canvasCreated) {
      this._createCanvas();
    } else if (!this._ensureCanvasInContainer()) {
      this._createCanvas();
    }
    if (!this._imageData || !this._ctx || !this._pixelTemplate) return;
    const data = this._imageData.data;
    const scale = this.scale;
    const canvasWidth = this.width * scale;
    const template = this._pixelTemplate;
    const bgR = 10, bgG = 10, bgB = 10;
    for (let ledY = 0; ledY < this.height; ledY++) {
      for (let ledX = 0; ledX < this.width; ledX++) {
        const bufferIdx = ledY * this.width + ledX;
        const color = this.buffer[bufferIdx];
        if (!color || !Array.isArray(color)) continue;
        const r = Math.round(color[0]);
        const g = Math.round(color[1]);
        const b = Math.round(color[2]);
        const baseX = ledX * scale;
        const baseY = ledY * scale;
        for (let py = 0; py < scale; py++) {
          for (let px = 0; px < scale; px++) {
            const templateIdx = py * scale + px;
            const canvasIdx = ((baseY + py) * canvasWidth + (baseX + px)) * 4;
            if (template[templateIdx]) {
              data[canvasIdx] = r;
              data[canvasIdx + 1] = g;
              data[canvasIdx + 2] = b;
              data[canvasIdx + 3] = 255;
            } else {
              data[canvasIdx] = bgR;
              data[canvasIdx + 1] = bgG;
              data[canvasIdx + 2] = bgB;
              data[canvasIdx + 3] = 255;
            }
          }
        }
      }
    }
    this._ctx.putImageData(this._imageData, 0, 0);
    if (this.glowEnabled && this._glowCtx) {
      this._glowCtx.drawImage(this._canvas, 0, 0);
    }
  }
  setData(pixels, extendedPixels = null, extendedWidth = null) {
    this._colorPixels = pixels || [];
    if (extendedPixels) {
      this._extendedColorPixels = extendedPixels;
      this.extendedWidth = extendedWidth || this.width;
    } else {
      this._extendedColorPixels = pixels || [];
      this.extendedWidth = this.width;
    }
  }
  setEffect(effect, speed = 50) {
    const wasRunning = this._isRunning;
    if (this.effect !== effect) {
      this.effect = effect;
      this.effectManager.initEffect(effect, { speed });
    }
    this.speed = speed;
    if (wasRunning && effect !== "fixed") {
      this.start();
    }
  }
  start() {
    if (this._isRunning) return;
    this._isRunning = true;
    this.lastFrameTime = performance.now();
    this._animate();
  }
  stop() {
    this._isRunning = false;
    if (this.animationId) {
      cancelAnimationFrame(this.animationId);
      this.animationId = null;
    }
  }
  get isRunning() {
    return this._isRunning;
  }
  _animate() {
    if (!this._isRunning) return;
    const now = performance.now();
    const frameInterval = 500 - (this.speed - 1) * 4.7;
    if (now - this.lastFrameTime >= frameInterval) {
      this.lastFrameTime = now;
      this.effectManager.step();
    }
    this._renderFrame();
    this.animationId = requestAnimationFrame(() => this._animate());
  }
  _renderFrame() {
    this.effectManager.render(this._colorPixels, this._extendedColorPixels, this.extendedWidth);
    this.flush();
  }
  renderStatic() {
    if (!this._canvasCreated) this._createCanvas();
    this._renderFrame();
  }
  setDimensions(width, height) {
    if (width !== this.width || height !== this.height) {
      this.width = width;
      this.height = height;
      this.extendedWidth = width;
      this._initBuffer();
      this._canvasCreated = false;
      this.effectManager = new EffectManager(this);
      if (this.effect !== "fixed") {
        this.effectManager.initEffect(this.effect, { speed: this.speed });
      }
    }
  }
  setContainer(container) {
    if (container !== this.container) {
      this.container = container;
      if (this._wrapper && container) {
        container.innerHTML = "";
        container.appendChild(this._wrapper);
      }
    }
  }
  destroy() {
    this.stop();
    this._canvas = null;
    this._ctx = null;
    this._imageData = null;
    this._glowCanvas = null;
    this._glowCtx = null;
    this._wrapper = null;
    this._canvasCreated = false;
    this._pixelTemplate = null;
  }
}
class CanvasLEDRenderer {
  constructor(container, options = {}) {
    this.container = container;
    this.width = options.width || 64;
    this.height = options.height || 16;
    this.pixelGap = options.pixelGap || 0.15;
    this.glowEnabled = options.glow !== false;
    this.scale = options.scale || 8;
    this.buffer = [];
    this._initBuffer();
    this._colorPixels = [];
    this._extendedColorPixels = [];
    this.extendedWidth = this.width;
    this.effect = "fixed";
    this.speed = 50;
    this.animationId = null;
    this.lastFrameTime = 0;
    this._isRunning = false;
    this._canvas = null;
    this._ctx = null;
    this._glowCanvas = null;
    this._glowCtx = null;
    this._wrapper = null;
    this._canvasCreated = false;
    this.effectManager = new EffectManager(this);
  }
  _initBuffer() {
    this.buffer = [];
    for (let i = 0; i < this.width * this.height; i++) {
      this.buffer.push([0, 0, 0]);
    }
  }
  _createCanvas() {
    if (typeof document === "undefined") return;
    const canvasWidth = this.width * this.scale;
    const canvasHeight = this.height * this.scale;
    this._wrapper = document.createElement("div");
    this._wrapper.style.cssText = `
      position: relative;
      width: 100%;
      aspect-ratio: ${this.width} / ${this.height};
      background: #0a0a0a;
      border-radius: 4px;
      overflow: hidden;
    `;
    if (this.glowEnabled) {
      this._glowCanvas = document.createElement("canvas");
      this._glowCanvas.width = canvasWidth;
      this._glowCanvas.height = canvasHeight;
      this._glowCanvas.style.cssText = `
        position: absolute; top: 0; left: 0; width: 100%; height: 100%;
        filter: blur(${this.scale * 0.8}px); opacity: 0.6;
        image-rendering: pixelated;
      `;
      this._glowCtx = this._glowCanvas.getContext("2d", { alpha: false });
      this._wrapper.appendChild(this._glowCanvas);
    }
    this._canvas = document.createElement("canvas");
    this._canvas.width = canvasWidth;
    this._canvas.height = canvasHeight;
    this._canvas.style.cssText = `
      position: absolute; top: 0; left: 0; width: 100%; height: 100%;
      image-rendering: pixelated; image-rendering: crisp-edges;
    `;
    this._ctx = this._canvas.getContext("2d", { alpha: false });
    if (this._ctx) this._ctx.imageSmoothingEnabled = false;
    this._wrapper.appendChild(this._canvas);
    if (this.container && this.container.isConnected !== false) {
      this.container.innerHTML = "";
      this.container.appendChild(this._wrapper);
    }
    this._canvasCreated = true;
  }
  _ensureCanvasInContainer() {
    if (!this.container) return false;
    if (this._wrapper && this._wrapper.parentNode === this.container) return true;
    if (this._wrapper && this.container.isConnected !== false) {
      this.container.innerHTML = "";
      this.container.appendChild(this._wrapper);
      return true;
    }
    return false;
  }
  setPixel(x, y, color) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      const idx = y * this.width + x;
      if (idx < this.buffer.length) {
        this.buffer[idx] = color;
      }
    }
  }
  clear() {
    for (let i = 0; i < this.buffer.length; i++) {
      this.buffer[i] = [0, 0, 0];
    }
  }
  flush() {
    if (!this._canvasCreated) {
      this._createCanvas();
    } else if (!this._ensureCanvasInContainer()) {
      this._createCanvas();
    }
    const ctx = this._ctx;
    if (!ctx) return;
    const scale = this.scale;
    const gap = Math.max(1, Math.floor(scale * this.pixelGap));
    const pixelSize = scale - gap;
    const cornerRadius = Math.max(1, Math.floor(scale * 0.15));
    ctx.fillStyle = "#0a0a0a";
    ctx.fillRect(0, 0, this._canvas.width, this._canvas.height);
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const idx = y * this.width + x;
        const color = this.buffer[idx];
        if (!color || !Array.isArray(color)) continue;
        const [r, g, b] = color;
        const px = x * scale;
        const py = y * scale;
        ctx.fillStyle = `rgb(${Math.round(r)}, ${Math.round(g)}, ${Math.round(b)})`;
        this._roundRect(ctx, px, py, pixelSize, pixelSize, cornerRadius);
      }
    }
    if (this.glowEnabled && this._glowCtx) {
      this._glowCtx.drawImage(this._canvas, 0, 0);
    }
  }
  _roundRect(ctx, x, y, w, h, r) {
    ctx.beginPath();
    ctx.moveTo(x + r, y);
    ctx.lineTo(x + w - r, y);
    ctx.quadraticCurveTo(x + w, y, x + w, y + r);
    ctx.lineTo(x + w, y + h - r);
    ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
    ctx.lineTo(x + r, y + h);
    ctx.quadraticCurveTo(x, y + h, x, y + h - r);
    ctx.lineTo(x, y + r);
    ctx.quadraticCurveTo(x, y, x + r, y);
    ctx.closePath();
    ctx.fill();
  }
  setData(pixels, extendedPixels = null, extendedWidth = null) {
    this._colorPixels = pixels || [];
    if (extendedPixels) {
      this._extendedColorPixels = extendedPixels;
      this.extendedWidth = extendedWidth || this.width;
    } else {
      this._extendedColorPixels = pixels || [];
      this.extendedWidth = this.width;
    }
  }
  setEffect(effect, speed = 50) {
    const wasRunning = this._isRunning;
    if (this.effect !== effect) {
      this.effect = effect;
      this.effectManager.initEffect(effect, { speed });
    }
    this.speed = speed;
    if (wasRunning && effect !== "fixed") {
      this.start();
    }
  }
  start() {
    if (this._isRunning) return;
    this._isRunning = true;
    this.lastFrameTime = performance.now();
    this._animate();
  }
  stop() {
    this._isRunning = false;
    if (this.animationId) {
      cancelAnimationFrame(this.animationId);
      this.animationId = null;
    }
  }
  get isRunning() {
    return this._isRunning;
  }
  _animate() {
    if (!this._isRunning) return;
    const now = performance.now();
    const frameInterval = 500 - (this.speed - 1) * 4.7;
    if (now - this.lastFrameTime >= frameInterval) {
      this.lastFrameTime = now;
      this.effectManager.step();
    }
    this._renderFrame();
    this.animationId = requestAnimationFrame(() => this._animate());
  }
  _renderFrame() {
    this.effectManager.render(this._colorPixels, this._extendedColorPixels, this.extendedWidth);
    this.flush();
  }
  renderStatic() {
    if (!this._canvasCreated) this._createCanvas();
    this._renderFrame();
  }
  setDimensions(width, height) {
    if (width !== this.width || height !== this.height) {
      this.width = width;
      this.height = height;
      this.extendedWidth = width;
      this._initBuffer();
      this._canvasCreated = false;
      this.effectManager = new EffectManager(this);
      if (this.effect !== "fixed") {
        this.effectManager.initEffect(this.effect, { speed: this.speed });
      }
    }
  }
  setContainer(container) {
    if (container !== this.container) {
      this.container = container;
      if (this._wrapper && container) {
        container.innerHTML = "";
        container.appendChild(this._wrapper);
      }
    }
  }
  destroy() {
    this.stop();
    this._canvas = null;
    this._ctx = null;
    this._glowCanvas = null;
    this._glowCtx = null;
    this._wrapper = null;
    this._canvasCreated = false;
  }
}
class SVGLEDMatrixRenderer {
  constructor(container, options = {}) {
    this.container = container;
    this.width = options.width || 64;
    this.height = options.height || 16;
    this.pixelGap = options.pixelGap || 0.1;
    this.buffer = [];
    this.prevBuffer = [];
    this._initBuffer();
    this._colorPixels = [];
    this._extendedColorPixels = [];
    this.extendedWidth = this.width;
    this.effect = "fixed";
    this.speed = 50;
    this.animationId = null;
    this.lastFrameTime = 0;
    this._isRunning = false;
    this.pixelElements = [];
    this.svgCreated = false;
    this._svg = null;
    this.effectManager = new EffectManager(this);
  }
  _initBuffer() {
    this.buffer = [];
    this.prevBuffer = [];
    for (let i = 0; i < this.width * this.height; i++) {
      this.buffer.push([0, 0, 0]);
      this.prevBuffer.push([-1, -1, -1]);
    }
  }
  _createSvg() {
    if (typeof document === "undefined") return;
    const svgWidth = 100;
    const pxWidth = svgWidth / this.width;
    const pxHeight = pxWidth;
    const svgHeight = this.height * pxHeight;
    const gap = this.pixelGap;
    const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
    svg.setAttribute("viewBox", `0 0 ${svgWidth} ${svgHeight}`);
    svg.setAttribute("preserveAspectRatio", "xMidYMid meet");
    svg.style.width = "100%";
    svg.style.height = "100%";
    svg.style.display = "block";
    this.pixelElements = [];
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const rect = document.createElementNS("http://www.w3.org/2000/svg", "rect");
        rect.setAttribute("x", String(x * pxWidth));
        rect.setAttribute("y", String(y * pxHeight));
        rect.setAttribute("width", String(pxWidth - gap));
        rect.setAttribute("height", String(pxHeight - gap));
        rect.setAttribute("rx", "0.3");
        rect.setAttribute("fill", "rgb(17, 17, 17)");
        svg.appendChild(rect);
        this.pixelElements.push(rect);
      }
    }
    if (this.container && this.container.isConnected !== false) {
      this.container.innerHTML = "";
      this.container.appendChild(svg);
    }
    this._svg = svg;
    this.svgCreated = true;
  }
  _ensureSvgInContainer() {
    if (!this.container) return false;
    if (this._svg && this._svg.parentNode === this.container) return true;
    if (this._svg && this.container.isConnected !== false) {
      this.container.innerHTML = "";
      this.container.appendChild(this._svg);
      return true;
    }
    return false;
  }
  setPixel(x, y, color) {
    if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
      const idx = y * this.width + x;
      if (idx < this.buffer.length) {
        this.buffer[idx] = color;
      }
    }
  }
  clear() {
    for (let i = 0; i < this.buffer.length; i++) {
      this.buffer[i] = [0, 0, 0];
    }
  }
  flush() {
    if (!this.svgCreated) {
      this._createSvg();
    } else if (!this._ensureSvgInContainer()) {
      this._createSvg();
    }
    for (let i = 0; i < this.buffer.length; i++) {
      const bufferItem = this.buffer[i];
      const prevItem = this.prevBuffer[i];
      if (!bufferItem || !Array.isArray(bufferItem)) continue;
      if (!prevItem || !Array.isArray(prevItem)) {
        this.prevBuffer[i] = [-1, -1, -1];
        continue;
      }
      const [r, g, b] = bufferItem;
      const [pr, pg, pb] = prevItem;
      if (r !== pr || g !== pg || b !== pb) {
        const rect = this.pixelElements[i];
        if (rect) {
          const isLit = r > 20 || g > 20 || b > 20;
          rect.setAttribute("fill", `rgb(${Math.round(r)}, ${Math.round(g)}, ${Math.round(b)})`);
          if (isLit) {
            rect.style.filter = `drop-shadow(0 0 2px rgb(${Math.round(r)}, ${Math.round(g)}, ${Math.round(b)}))`;
          } else {
            rect.style.filter = "";
          }
        }
        this.prevBuffer[i] = [r, g, b];
      }
    }
  }
  setData(pixels, extendedPixels = null, extendedWidth = null) {
    this._colorPixels = pixels || [];
    if (extendedPixels) {
      this._extendedColorPixels = extendedPixels;
      this.extendedWidth = extendedWidth || this.width;
    } else {
      this._extendedColorPixels = pixels || [];
      this.extendedWidth = this.width;
    }
  }
  setEffect(effect, speed = 50) {
    const wasRunning = this._isRunning;
    if (this.effect !== effect) {
      this.effect = effect;
      this.effectManager.initEffect(effect, { speed });
    }
    this.speed = speed;
    if (wasRunning && effect !== "fixed") {
      this.start();
    }
  }
  start() {
    if (this._isRunning) return;
    this._isRunning = true;
    this.lastFrameTime = performance.now();
    this._animate();
  }
  stop() {
    this._isRunning = false;
    if (this.animationId) {
      cancelAnimationFrame(this.animationId);
      this.animationId = null;
    }
  }
  get isRunning() {
    return this._isRunning;
  }
  _animate() {
    if (!this._isRunning) return;
    const now = performance.now();
    const frameInterval = 500 - (this.speed - 1) * 4.7;
    if (now - this.lastFrameTime >= frameInterval) {
      this.lastFrameTime = now;
      this.effectManager.step();
    }
    this._renderFrame();
    this.animationId = requestAnimationFrame(() => this._animate());
  }
  _renderFrame() {
    this.effectManager.render(this._colorPixels, this._extendedColorPixels, this.extendedWidth);
    this.flush();
  }
  renderStatic() {
    if (!this.svgCreated) this._createSvg();
    this._renderFrame();
  }
  setDimensions(width, height) {
    if (width !== this.width || height !== this.height) {
      this.width = width;
      this.height = height;
      this.extendedWidth = width;
      this._initBuffer();
      this.svgCreated = false;
      this.effectManager = new EffectManager(this);
      if (this.effect !== "fixed") {
        this.effectManager.initEffect(this.effect, { speed: this.speed });
      }
    }
  }
  setContainer(container) {
    if (container !== this.container) {
      this.container = container;
      if (this._svg && container) {
        container.innerHTML = "";
        container.appendChild(this._svg);
      }
    }
  }
  destroy() {
    this.stop();
    this.pixelElements = [];
    this._svg = null;
    this.svgCreated = false;
  }
}
function createPixelSvg(width, height, pixels, pixelGap = 1) {
  const svgWidth = 100;
  const pxWidth = svgWidth / width;
  const pxHeight = pxWidth;
  const svgHeight = height * pxHeight;
  const gap = pixelGap * 0.1;
  let rects = "";
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const color = pixels[y * width + x] || "#111";
      const isLit = color !== "#111" && color !== "#000" && color !== "#1a1a1a" && color !== "#050505";
      const style = isLit ? `filter:drop-shadow(0 0 2px ${color});` : "";
      rects += `<rect x="${x * pxWidth}" y="${y * pxHeight}" width="${pxWidth - gap}" height="${pxHeight - gap}" fill="${color}" rx="0.3" style="${style}"/>`;
    }
  }
  return `<svg viewBox="0 0 ${svgWidth} ${svgHeight}" preserveAspectRatio="xMidYMid meet" style="width:100%;height:100%;display:block;">${rects}</svg>`;
}
const pixelFont = {
  "A": [124, 18, 17, 18, 124],
  "B": [127, 73, 73, 73, 54],
  "C": [62, 65, 65, 65, 34],
  "D": [127, 65, 65, 34, 28],
  "E": [127, 73, 73, 73, 65],
  "F": [127, 9, 9, 9, 1],
  "G": [62, 65, 73, 73, 122],
  "H": [127, 8, 8, 8, 127],
  "I": [0, 65, 127, 65, 0],
  "J": [32, 64, 65, 63, 1],
  "K": [127, 8, 20, 34, 65],
  "L": [127, 64, 64, 64, 64],
  "M": [127, 2, 12, 2, 127],
  "N": [127, 4, 8, 16, 127],
  "O": [62, 65, 65, 65, 62],
  "P": [127, 9, 9, 9, 6],
  "Q": [62, 65, 81, 33, 94],
  "R": [127, 9, 25, 41, 70],
  "S": [70, 73, 73, 73, 49],
  "T": [1, 1, 127, 1, 1],
  "U": [63, 64, 64, 64, 63],
  "V": [31, 32, 64, 32, 31],
  "W": [63, 64, 56, 64, 63],
  "X": [99, 20, 8, 20, 99],
  "Y": [7, 8, 112, 8, 7],
  "Z": [97, 81, 73, 69, 67],
  "a": [32, 84, 84, 84, 120],
  "b": [127, 72, 68, 68, 56],
  "c": [56, 68, 68, 68, 32],
  "d": [56, 68, 68, 72, 127],
  "e": [56, 84, 84, 84, 24],
  "f": [8, 126, 9, 1, 2],
  "g": [12, 82, 82, 82, 62],
  "h": [127, 8, 4, 4, 120],
  "i": [0, 68, 125, 64, 0],
  "j": [32, 64, 68, 61, 0],
  "k": [127, 16, 40, 68, 0],
  "l": [0, 65, 127, 64, 0],
  "m": [124, 4, 24, 4, 120],
  "n": [124, 8, 4, 4, 120],
  "o": [56, 68, 68, 68, 56],
  "p": [124, 20, 20, 20, 8],
  "q": [8, 20, 20, 24, 124],
  "r": [124, 8, 4, 4, 8],
  "s": [72, 84, 84, 84, 32],
  "t": [4, 63, 68, 64, 32],
  "u": [60, 64, 64, 32, 124],
  "v": [28, 32, 64, 32, 28],
  "w": [60, 64, 48, 64, 60],
  "x": [68, 40, 16, 40, 68],
  "y": [12, 80, 80, 80, 60],
  "z": [68, 100, 84, 76, 68],
  "0": [62, 81, 73, 69, 62],
  "1": [0, 66, 127, 64, 0],
  "2": [66, 97, 81, 73, 70],
  "3": [33, 65, 69, 75, 49],
  "4": [24, 20, 18, 127, 16],
  "5": [39, 69, 69, 69, 57],
  "6": [60, 74, 73, 73, 48],
  "7": [1, 113, 9, 5, 3],
  "8": [54, 73, 73, 73, 54],
  "9": [6, 73, 73, 41, 30],
  " ": [0, 0, 0, 0, 0],
  ".": [0, 96, 96, 0, 0],
  ",": [0, 128, 96, 0, 0],
  ":": [0, 54, 54, 0, 0],
  ";": [0, 128, 54, 0, 0],
  "!": [0, 0, 95, 0, 0],
  "?": [2, 1, 81, 9, 6],
  "-": [8, 8, 8, 8, 8],
  "+": [8, 8, 62, 8, 8],
  "=": [20, 20, 20, 20, 20],
  "_": [64, 64, 64, 64, 64],
  "/": [32, 16, 8, 4, 2],
  "\\": [2, 4, 8, 16, 32],
  "(": [0, 28, 34, 65, 0],
  ")": [0, 65, 34, 28, 0],
  "[": [0, 127, 65, 65, 0],
  "]": [0, 65, 65, 127, 0],
  "<": [8, 20, 34, 65, 0],
  ">": [0, 65, 34, 20, 8],
  "*": [20, 8, 62, 8, 20],
  "#": [20, 127, 20, 127, 20],
  "@": [62, 65, 93, 85, 30],
  "&": [54, 73, 85, 34, 80],
  "%": [35, 19, 8, 100, 98],
  "$": [18, 42, 127, 42, 36],
  "'": [0, 0, 7, 0, 0],
  '"': [0, 7, 0, 7, 0],
  "`": [0, 1, 2, 0, 0],
  "^": [4, 2, 1, 2, 4],
  "~": [8, 4, 8, 16, 8]
};
function textToPixels(text, width, height, fgColor = "#ff6600", bgColor = "#111") {
  const pixels = [];
  const charWidth = 6;
  const charHeight = 7;
  const startY = Math.floor((height - charHeight) / 2);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      pixels.push(bgColor);
    }
  }
  const textWidth = text.length * charWidth - 1;
  const startX = Math.max(1, Math.floor((width - textWidth) / 2));
  let xOffset = startX;
  for (const char of text) {
    const charData = pixelFont[char] || pixelFont[" "];
    for (let col = 0; col < 5; col++) {
      for (let row = 0; row < 7; row++) {
        const pixelOn = charData[col] >> row & 1;
        const px = xOffset + col;
        const py = startY + row;
        if (px >= 0 && px < width && py < height && py >= 0) {
          pixels[py * width + px] = pixelOn ? fgColor : bgColor;
        }
      }
    }
    xOffset += charWidth;
  }
  return pixels;
}
function textToScrollPixels(text, displayWidth, height, fgColor = "#ff6600", bgColor = "#111") {
  const charWidth = 6;
  const charHeight = 7;
  const startY = Math.floor((height - charHeight) / 2);
  const textPixelWidth = text.length * charWidth;
  const extendedWidth = displayWidth + textPixelWidth + displayWidth;
  const pixels = [];
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < extendedWidth; x++) {
      pixels.push(bgColor);
    }
  }
  let xOffset = displayWidth;
  for (const char of text) {
    const charData = pixelFont[char] || pixelFont[" "];
    for (let col = 0; col < 5; col++) {
      for (let row = 0; row < 7; row++) {
        const pixelOn = charData[col] >> row & 1;
        const px = xOffset + col;
        const py = startY + row;
        if (px >= 0 && px < extendedWidth && py < height && py >= 0) {
          pixels[py * extendedWidth + px] = pixelOn ? fgColor : bgColor;
        }
      }
    }
    xOffset += charWidth;
  }
  return { pixels, width: extendedWidth };
}
const FONT_METRICS = {
  "VCR_OSD_MONO": {
    16: { font_size: 16, offset: [0, 0], pixel_threshold: 70, var_width: true },
    24: { font_size: 24, offset: [0, 0], pixel_threshold: 70, var_width: true },
    32: { font_size: 28, offset: [-1, 2], pixel_threshold: 30, var_width: false }
  },
  "CUSONG": {
    16: { font_size: 16, offset: [0, -1], pixel_threshold: 70, var_width: false },
    24: { font_size: 24, offset: [0, 0], pixel_threshold: 70, var_width: false },
    32: { font_size: 32, offset: [0, 0], pixel_threshold: 70, var_width: false }
  }
};
const fontLoadState = {};
const fontLoadPromises$1 = {};
const defaultResolver$1 = (fontName) => {
  if (typeof window === "undefined") return `/fonts/${fontName}.ttf`;
  const basePath = window.location.pathname.substring(
    0,
    window.location.pathname.lastIndexOf("/") + 1
  );
  return `${basePath}fonts/${fontName}.ttf`;
};
let _fontResolver$1 = defaultResolver$1;
function setFontResolver$1(resolver) {
  _fontResolver$1 = resolver;
}
function getHeightKey$1(height) {
  if (height <= 18) return 16;
  if (height <= 28) return 24;
  return 32;
}
async function loadFont(fontName, resolver) {
  if (fontLoadState[fontName] === true) return true;
  if (fontLoadState[fontName] === false) return false;
  if (fontName in fontLoadPromises$1) return fontLoadPromises$1[fontName];
  fontLoadPromises$1[fontName] = (async () => {
    if (typeof document === "undefined") return false;
    const resolveUrl = resolver || _fontResolver$1;
    const fontUrl = resolveUrl(fontName);
    try {
      const font = new FontFace(fontName, `url(${fontUrl})`);
      const loadedFont = await font.load();
      document.fonts.add(loadedFont);
      fontLoadState[fontName] = true;
      return true;
    } catch (e) {
      console.warn(`PixelDisplay: Failed to load font ${fontName}:`, e);
      fontLoadState[fontName] = false;
      return false;
    }
  })();
  return fontLoadPromises$1[fontName];
}
function isFontLoaded(fontName) {
  return fontLoadState[fontName] === true;
}
function textToPixelsCanvas(text, width, height, fgColor = "#ff6600", bgColor = "#111", fontName = "VCR_OSD_MONO") {
  if (typeof document === "undefined") return null;
  const fontMetrics = FONT_METRICS[fontName];
  if (!fontMetrics) return null;
  if (!isFontLoaded(fontName)) {
    loadFont(fontName);
    return null;
  }
  const heightKey = getHeightKey$1(height);
  const metrics = fontMetrics[heightKey];
  const canvas = document.createElement("canvas");
  canvas.width = width;
  canvas.height = height;
  const ctx = canvas.getContext("2d");
  if (!ctx) return null;
  ctx.imageSmoothingEnabled = false;
  ctx.fillStyle = bgColor;
  ctx.fillRect(0, 0, width, height);
  if (!text || text.trim() === "") {
    const pixels2 = [];
    for (let i = 0; i < width * height; i++) pixels2.push(bgColor);
    return pixels2;
  }
  ctx.font = `${metrics.font_size}px "${fontName}"`;
  ctx.fillStyle = fgColor;
  ctx.textBaseline = "top";
  const textMetrics = ctx.measureText(text);
  const textWidth = textMetrics.width;
  const x = Math.floor((width - textWidth) / 2) + metrics.offset[0];
  const y = Math.floor((height - metrics.font_size) / 2) + metrics.offset[1];
  ctx.fillText(text, x, y);
  const imageData = ctx.getImageData(0, 0, width, height);
  const pixels = [];
  for (let i = 0; i < imageData.data.length; i += 4) {
    const r = imageData.data[i];
    const g = imageData.data[i + 1];
    const b = imageData.data[i + 2];
    const gray = (r + g + b) / 3;
    pixels.push(gray >= metrics.pixel_threshold ? fgColor : bgColor);
  }
  return pixels;
}
function textToScrollPixelsCanvas(text, displayWidth, height, fgColor = "#ff6600", bgColor = "#111", fontName = "VCR_OSD_MONO") {
  if (typeof document === "undefined") return null;
  const fontMetrics = FONT_METRICS[fontName];
  if (!fontMetrics) return null;
  if (!isFontLoaded(fontName)) {
    loadFont(fontName);
    return null;
  }
  const heightKey = getHeightKey$1(height);
  const metrics = fontMetrics[heightKey];
  const tempCanvas = document.createElement("canvas");
  const tempCtx = tempCanvas.getContext("2d");
  if (!tempCtx) return null;
  tempCtx.font = `${metrics.font_size}px "${fontName}"`;
  const textWidth = Math.ceil(tempCtx.measureText(text).width);
  const extendedWidth = displayWidth + textWidth + displayWidth;
  const canvas = document.createElement("canvas");
  canvas.width = extendedWidth;
  canvas.height = height;
  const ctx = canvas.getContext("2d");
  if (!ctx) return null;
  ctx.imageSmoothingEnabled = false;
  ctx.fillStyle = bgColor;
  ctx.fillRect(0, 0, extendedWidth, height);
  if (!text || text.trim() === "") {
    const pixels2 = [];
    for (let i = 0; i < extendedWidth * height; i++) pixels2.push(bgColor);
    return { pixels: pixels2, width: extendedWidth };
  }
  ctx.font = `${metrics.font_size}px "${fontName}"`;
  ctx.fillStyle = fgColor;
  ctx.textBaseline = "top";
  const x = displayWidth + metrics.offset[0];
  const y = Math.floor((height - metrics.font_size) / 2) + metrics.offset[1];
  ctx.fillText(text, x, y);
  const imageData = ctx.getImageData(0, 0, extendedWidth, height);
  const pixels = [];
  for (let i = 0; i < imageData.data.length; i += 4) {
    const r = imageData.data[i];
    const g = imageData.data[i + 1];
    const b = imageData.data[i + 2];
    const gray = (r + g + b) / 3;
    pixels.push(gray >= metrics.pixel_threshold ? fgColor : bgColor);
  }
  return { pixels, width: extendedWidth };
}
async function preloadFonts(resolver) {
  const fonts = Object.keys(FONT_METRICS);
  await Promise.all(fonts.map((f) => loadFont(f, resolver)));
}
let $Font = null;
let $fetchline = null;
async function ensureBdfParser() {
  if ($Font && $fetchline) return true;
  try {
    const bdfparser = await import("./index-Ds5kwaZ_.js");
    const fetchline = await import("./index-Cl4FejWM.js");
    $Font = bdfparser.$Font;
    const fl = fetchline;
    $fetchline = fl.default || fl.$fetchline || fetchline;
    return true;
  } catch {
    console.warn("PixelDisplay: bdfparser/fetchline packages not available. BDF font rendering disabled.");
    return false;
  }
}
const BDF_FONT_CONFIG = {
  "VCR_OSD_MONO": {
    16: { file: "VCR_OSD_MONO_16.bdf", yOffset: 0 },
    24: { file: "VCR_OSD_MONO_24.bdf", yOffset: 0 },
    32: { file: "VCR_OSD_MONO_32.bdf", yOffset: 2 }
  },
  "CUSONG": {
    16: { file: "CUSONG_16.bdf", yOffset: -1 },
    24: { file: "CUSONG_24.bdf", yOffset: 0 },
    32: { file: "CUSONG_32.bdf", yOffset: 0 }
  }
};
const fontCache = /* @__PURE__ */ new Map();
const fontLoadPromises = /* @__PURE__ */ new Map();
const defaultResolver = (_fontName, fileName) => {
  if (typeof window === "undefined") return `/fonts/${fileName || _fontName}`;
  const basePath = window.location.pathname.substring(
    0,
    window.location.pathname.lastIndexOf("/") + 1
  );
  return `${basePath}fonts/${fileName || _fontName}`;
};
let _fontResolver = defaultResolver;
function setFontResolver(resolver) {
  _fontResolver = resolver;
}
function getHeightKey(height) {
  if (height <= 18) return 16;
  if (height <= 28) return 24;
  return 32;
}
async function loadBdfFont(fontName, heightKey = 16, resolver) {
  const cacheKey = `${fontName}_${heightKey}`;
  if (fontCache.has(cacheKey)) {
    return fontCache.get(cacheKey);
  }
  if (fontLoadPromises.has(cacheKey)) {
    return fontLoadPromises.get(cacheKey);
  }
  const fontConfig = BDF_FONT_CONFIG[fontName];
  if (!fontConfig || !fontConfig[heightKey]) {
    console.warn(`PixelDisplay BDF: No config for font ${fontName} at height ${heightKey}`);
    return null;
  }
  const config = fontConfig[heightKey];
  const loadPromise = (async () => {
    try {
      const parserAvailable = await ensureBdfParser();
      if (!parserAvailable || !$Font || !$fetchline) return null;
      const resolveUrl = resolver || _fontResolver;
      const fontUrl = resolveUrl(fontName, config.file);
      const font = await $Font($fetchline(fontUrl));
      const result = { font, config };
      fontCache.set(cacheKey, result);
      return result;
    } catch (e) {
      console.warn(`PixelDisplay BDF: Failed to load font ${fontName} (${heightKey}px):`, e);
      fontLoadPromises.delete(cacheKey);
      return null;
    }
  })();
  fontLoadPromises.set(cacheKey, loadPromise);
  return loadPromise;
}
function isBdfFontLoaded(fontName, heightKey = 16) {
  const cacheKey = `${fontName}_${heightKey}`;
  return fontCache.has(cacheKey);
}
function textToPixelsBdf(text, width, height, fgColor = "#ff6600", bgColor = "#111", fontName = "VCR_OSD_MONO") {
  const heightKey = getHeightKey(height);
  const cacheKey = `${fontName}_${heightKey}`;
  const cached = fontCache.get(cacheKey);
  if (!cached) {
    loadBdfFont(fontName, heightKey);
    return null;
  }
  const { font, config } = cached;
  const pixels = new Array(width * height).fill(bgColor);
  if (!text || text.trim() === "") return pixels;
  try {
    const bitmap = font.draw(text, { direction: "lrtb", mode: 1 });
    const bindata = bitmap.bindata;
    const textWidth = bitmap.width();
    const textHeight = bitmap.height();
    const xOffset = Math.floor((width - textWidth) / 2);
    const yOffset = Math.floor((height - textHeight) / 2) + (config.yOffset || 0);
    for (let row = 0; row < textHeight; row++) {
      const rowData = bindata[row] || "";
      for (let col = 0; col < rowData.length; col++) {
        const px = xOffset + col;
        const py = yOffset + row;
        if (px >= 0 && px < width && py >= 0 && py < height) {
          const idx = py * width + px;
          pixels[idx] = rowData[col] === "1" ? fgColor : bgColor;
        }
      }
    }
  } catch (e) {
    console.warn("PixelDisplay BDF: Error rendering text:", e);
    return null;
  }
  return pixels;
}
function textToScrollPixelsBdf(text, displayWidth, height, fgColor = "#ff6600", bgColor = "#111", fontName = "VCR_OSD_MONO") {
  const heightKey = getHeightKey(height);
  const cacheKey = `${fontName}_${heightKey}`;
  const cached = fontCache.get(cacheKey);
  if (!cached) {
    loadBdfFont(fontName, heightKey);
    return null;
  }
  const { font, config } = cached;
  if (!text || text.trim() === "") {
    const extendedWidth = displayWidth * 3;
    const pixels = new Array(extendedWidth * height).fill(bgColor);
    return { pixels, width: extendedWidth };
  }
  try {
    const bitmap = font.draw(text, { direction: "lrtb", mode: 1 });
    const bindata = bitmap.bindata;
    const textWidth = bitmap.width();
    const textHeight = bitmap.height();
    const extendedWidth = displayWidth + textWidth + displayWidth;
    const pixels = new Array(extendedWidth * height).fill(bgColor);
    const xStart = displayWidth;
    const yOffset = Math.floor((height - textHeight) / 2) + (config.yOffset || 0);
    for (let row = 0; row < textHeight; row++) {
      const rowData = bindata[row] || "";
      for (let col = 0; col < rowData.length; col++) {
        const px = xStart + col;
        const py = yOffset + row;
        if (px >= 0 && px < extendedWidth && py >= 0 && py < height) {
          const idx = py * extendedWidth + px;
          pixels[idx] = rowData[col] === "1" ? fgColor : bgColor;
        }
      }
    }
    return { pixels, width: extendedWidth };
  } catch (e) {
    console.warn("PixelDisplay BDF: Error rendering scroll text:", e);
    return null;
  }
}
async function preloadBdfFonts(resolver) {
  const fonts = Object.keys(BDF_FONT_CONFIG);
  const heights = [16, 24, 32];
  const promises = [];
  for (const fontName of fonts) {
    for (const height of heights) {
      if (BDF_FONT_CONFIG[fontName][height]) {
        promises.push(loadBdfFont(fontName, height, resolver));
      }
    }
  }
  await Promise.all(promises);
}
function getAvailableBdfFonts() {
  return Object.keys(BDF_FONT_CONFIG);
}
function configureFonts(options) {
  if (options.baseUrl) {
    const base = options.baseUrl.replace(/\/+$/, "");
    setFontResolver$1((name) => `${base}/${name}.ttf`);
    setFontResolver((_name, file) => `${base}/${file || _name}`);
  }
  if (options.ttfResolver) setFontResolver$1(options.ttfResolver);
  if (options.bdfResolver) setFontResolver(options.bdfResolver);
}
export {
  ALL_EFFECTS as A,
  BDF_FONT_CONFIG as B,
  CanvasLEDRenderer as C,
  preloadBdfFonts as D,
  EFFECTS as E,
  FONT_METRICS as F,
  preloadFonts as G,
  setFontResolver as H,
  ImageDataLEDRenderer as I,
  setFontResolver$1 as J,
  SVGLEDMatrixRenderer as S,
  TEXT_EFFECTS as T,
  isFontLoaded as a,
  loadFont as b,
  textToPixelsCanvas as c,
  textToPixels as d,
  textToScrollPixelsBdf as e,
  textToScrollPixelsCanvas as f,
  textToScrollPixels as g,
  getHeightKey as h,
  isBdfFontLoaded as i,
  AMBIENT_EFFECTS as j,
  AmbientEffects as k,
  loadBdfFont as l,
  COLOR_EFFECTS as m,
  ColorEffects as n,
  EFFECT_CATEGORIES as o,
  EffectManager as p,
  TextEffects as q,
  configureFonts as r,
  createPixelSvg as s,
  textToPixelsBdf as t,
  getAvailableBdfFonts as u,
  getHeightKey$1 as v,
  hexToRgb as w,
  hexToRgbObj as x,
  hsvToRgb as y,
  pixelFont as z
};
