// react-pixel-display/dist/index.js
import { jsx } from "react/jsx-runtime";
import { useRef, useEffect, useCallback, useState, useMemo, forwardRef, useImperativeHandle } from "react";
import { C as CanvasLEDRenderer, S as SVGLEDMatrixRenderer, I as ImageDataLEDRenderer, i as isBdfFontLoaded, l as loadBdfFont, a as isFontLoaded, b as loadFont, t as textToPixelsBdf, c as textToPixelsCanvas, d as textToPixels, e as textToScrollPixelsBdf, f as textToScrollPixelsCanvas, g as textToScrollPixels, h as getHeightKey, E as EFFECTS } from "./index-DWEf46Kx.js";
import { A, j, k, B, m, n, o, p, F, T, q, r, s, u, v, w, x, y, z, D, G, H, J } from "./index-DWEf46Kx.js";
function usePixelDisplay(options) {
  const {
    width = 64,
    height = 16,
    renderer: type = "imagedata",
    glow = true,
    scale = 8,
    pixelGap = 0.15
  } = options;
  const containerRef = useRef(null);
  const rendererRef = useRef(null);
  useEffect(() => {
    var _a;
    const container = containerRef.current;
    if (!container) return;
    (_a = rendererRef.current) == null ? void 0 : _a.destroy();
    const opts = { width, height, glow, scale, pixelGap };
    if (type === "canvas") {
      rendererRef.current = new CanvasLEDRenderer(container, opts);
    } else if (type === "svg") {
      rendererRef.current = new SVGLEDMatrixRenderer(container, { width, height, pixelGap });
    } else {
      rendererRef.current = new ImageDataLEDRenderer(container, opts);
    }
    return () => {
      var _a2;
      (_a2 = rendererRef.current) == null ? void 0 : _a2.destroy();
      rendererRef.current = null;
    };
  }, [width, height, type, glow, scale, pixelGap]);
  const setData = useCallback(
    (pixels, extended, extW) => {
      var _a;
      (_a = rendererRef.current) == null ? void 0 : _a.setData(pixels, extended, extW);
    },
    []
  );
  const setEffect = useCallback((effect, speed) => {
    var _a;
    (_a = rendererRef.current) == null ? void 0 : _a.setEffect(effect, speed);
  }, []);
  const start = useCallback(() => {
    var _a;
    return (_a = rendererRef.current) == null ? void 0 : _a.start();
  }, []);
  const stop = useCallback(() => {
    var _a;
    return (_a = rendererRef.current) == null ? void 0 : _a.stop();
  }, []);
  const renderStatic = useCallback(() => {
    var _a;
    return (_a = rendererRef.current) == null ? void 0 : _a.renderStatic();
  }, []);
  return { containerRef, rendererRef, setData, setEffect, start, stop, renderStatic };
}
function useTextRenderer(options) {
  const {
    width,
    height,
    text,
    font = "legacy",
    fontName = "VCR_OSD_MONO",
    foregroundColor = "#ff6600",
    backgroundColor = "#111",
    fontResolver,
    needsScroll = false
  } = options;
  const [fontReady, setFontReady] = useState(font === "legacy");
  useEffect(() => {
    if (font === "legacy") {
      setFontReady(true);
      return;
    }
    const heightKey = getHeightKey(height);
    if (font === "bdf") {
      if (isBdfFontLoaded(fontName, heightKey)) {
        setFontReady(true);
      } else {
        setFontReady(false);
        loadBdfFont(fontName, heightKey, fontResolver).then((result) => {
          setFontReady(!!result);
        });
      }
    } else {
      if (isFontLoaded(fontName)) {
        setFontReady(true);
      } else {
        setFontReady(false);
        loadFont(fontName, fontResolver).then((result) => {
          setFontReady(result);
        });
      }
    }
  }, [font, fontName, height, fontResolver]);
  const { pixels, scrollData } = useMemo(() => {
    if (!fontReady) return { pixels: null, scrollData: null };
    let result = null;
    if (font === "bdf") {
      result = textToPixelsBdf(text, width, height, foregroundColor, backgroundColor, fontName);
    } else if (font === "canvas") {
      result = textToPixelsCanvas(text, width, height, foregroundColor, backgroundColor, fontName);
    }
    if (!result) {
      result = textToPixels(text, width, height, foregroundColor, backgroundColor);
    }
    let scroll = null;
    if (needsScroll) {
      if (font === "bdf") {
        scroll = textToScrollPixelsBdf(text, width, height, foregroundColor, backgroundColor, fontName);
      } else if (font === "canvas") {
        scroll = textToScrollPixelsCanvas(text, width, height, foregroundColor, backgroundColor, fontName);
      }
      if (!scroll) {
        scroll = textToScrollPixels(text, width, height, foregroundColor, backgroundColor);
      }
    }
    return { pixels: result, scrollData: scroll };
  }, [fontReady, width, height, text, font, fontName, foregroundColor, backgroundColor, needsScroll]);
  return { pixels, scrollData, fontReady };
}
const PixelDisplay = forwardRef(
  (props, ref) => {
    var _a;
    const {
      width = 64,
      height = 16,
      pixels: propPixels,
      text,
      font = "legacy",
      fontName = "VCR_OSD_MONO",
      foregroundColor = "#ff6600",
      backgroundColor = "#111",
      effect = "fixed",
      speed = 50,
      renderer: rendererType = "imagedata",
      glow = true,
      scale = 8,
      pixelGap = 0.15,
      fontResolver,
      onReady,
      className,
      style
    } = props;
    const isAmbient = ((_a = EFFECTS[effect]) == null ? void 0 : _a.category) === "ambient";
    const needsScroll = (effect === "scroll_ltr" || effect === "scroll_rtl" || effect === "bounce") && !!text && !isAmbient;
    const { containerRef, rendererRef, setData, setEffect, start, stop, renderStatic } = usePixelDisplay({ width, height, renderer: rendererType, glow, scale, pixelGap });
    const textOptions = useMemo(
      () => ({
        width,
        height,
        text: text || "",
        font,
        fontName,
        foregroundColor,
        backgroundColor,
        fontResolver,
        needsScroll
      }),
      [width, height, text, font, fontName, foregroundColor, backgroundColor, fontResolver, needsScroll]
    );
    const { pixels: textPixels, scrollData, fontReady } = useTextRenderer(textOptions);
    useEffect(() => {
      if (!rendererRef.current) return;
      if (isAmbient) {
        setData([], [], width);
      } else if (propPixels) {
        setData(propPixels);
      } else if (textPixels) {
        if (needsScroll && scrollData) {
          setData(textPixels, scrollData.pixels, scrollData.width);
        } else {
          setData(textPixels);
        }
      }
    }, [propPixels, textPixels, scrollData, isAmbient, needsScroll, width, setData, rendererRef]);
    useEffect(() => {
      if (!rendererRef.current) return;
      setEffect(effect, speed);
      if (effect === "fixed") {
        stop();
        renderStatic();
      } else {
        start();
      }
    }, [effect, speed, setEffect, start, stop, renderStatic, rendererRef]);
    useImperativeHandle(
      ref,
      () => ({
        start: () => {
          var _a2;
          return (_a2 = rendererRef.current) == null ? void 0 : _a2.start();
        },
        stop: () => {
          var _a2;
          return (_a2 = rendererRef.current) == null ? void 0 : _a2.stop();
        },
        getRenderer: () => rendererRef.current,
        isRunning: () => {
          var _a2;
          return ((_a2 = rendererRef.current) == null ? void 0 : _a2.isRunning) ?? false;
        }
      }),
      [rendererRef]
    );
    useEffect(() => {
      if (rendererRef.current && (propPixels || textPixels || isAmbient)) {
        onReady == null ? void 0 : onReady();
      }
    }, [rendererRef, propPixels, textPixels, isAmbient, onReady]);
    return /* @__PURE__ */ jsx(
      "div",
      {
        ref: containerRef,
        className,
        style: {
          background: "#0a0a0a",
          borderRadius: "4px",
          overflow: "hidden",
          ...style
        }
      }
    );
  }
);
PixelDisplay.displayName = "PixelDisplay";
export {
  A as ALL_EFFECTS,
  j as AMBIENT_EFFECTS,
  k as AmbientEffects,
  B as BDF_FONT_CONFIG,
  m as COLOR_EFFECTS,
  CanvasLEDRenderer,
  n as ColorEffects,
  EFFECTS,
  o as EFFECT_CATEGORIES,
  p as EffectManager,
  F as FONT_METRICS,
  ImageDataLEDRenderer,
  ImageDataLEDRenderer as LEDMatrixRenderer,
  PixelDisplay,
  SVGLEDMatrixRenderer,
  T as TEXT_EFFECTS,
  q as TextEffects,
  r as configureFonts,
  s as createPixelSvg,
  u as getAvailableBdfFonts,
  v as getHeightKey,
  w as hexToRgb,
  x as hexToRgbObj,
  y as hsvToRgb,
  isBdfFontLoaded,
  isFontLoaded,
  loadBdfFont,
  loadFont,
  z as pixelFont,
  D as preloadBdfFonts,
  G as preloadFonts,
  H as setBdfFontResolver,
  J as setTtfFontResolver,
  textToPixels,
  textToPixelsBdf,
  textToPixelsCanvas,
  textToScrollPixels,
  textToScrollPixelsBdf,
  textToScrollPixelsCanvas,
  usePixelDisplay,
  useTextRenderer
};
